design/one-pager-managed-resource-api-design.md

```diff
@@ -0,0 +1,294 @@
+* _Provider_. Cloud provider such as GCP, AWS, Azure offering IaaS, cloud networking, and managed services.
+* _Spec_. The sub-resource of Kubernetes resources that represents the desired state of the user.
+* _Status_ The sub-resource of Kubernetes resources that represents the most up-to-date information about the resource.
+
+For other terms, see [glossary].
+
+## Background
+Crossplane extends Kubernetes heavily and the maintainers always try to adopt the patterns that are already seen in 
+Kubernetes resources, notably declarative style API. However, it's not always intuitive for everyone to see what goes to
+where in a way that is consistent and just as expected by the consumers of those resources.
+As new controllers get implemented and new stacks are introduced by the community members, the need for consistent 
+pattern in our managed resources gets more important every day. This design document proposes some guiding principles 
+for designing managed resources as well as a few examples that show how these can map to actual resources.
+
+A few questions that this document needs to help answering:
+- What goes into `Spec` and what goes into `Status` of a managed resource?
+- Should _all_ information from the provider be represented in the Custom Resource?
+- What parts of the Custom Resource should be allowed to be consumed by the consumers?
+- Do we represent _desired_ state, _observed_ state or both?
+- How should we differentiate the fields that are used by provider from Crossplane's fields?
+
+This document tries to capture how the bridge between Crossplane and provider is shaped. However, each provider can
+wildly vary between each other. So, treat this document as a starting point, try to stick to it for consistency but
+deviate if required and open an issue about why you had to deviate so that we can reassess and improve it.
+
+## Guiding Principles
+
+Our users are the stakeholders on how we make decisions. There are two types of users for managed resources:
+* Input Users: Users who configure/create/manage the resource. Crossplane Claim Controller, Cluster Admin are a few
+  examples of this type. We want our `Spec` to be sole place for them to configure everything about the resource.
+* Output Users: Users who refer to/query/use/extract information from the managed resource. We want them to be able to
+  perform all read-only operations on the given resource as if they have the cloud provider's API available. This can
+  include configurations and metadata information about the resource. We also want them to access only `Status`
+  sub-resource of the managed resource because this is where the most up-to-date information about the resource is
+  stored and we are able to limit RBAC of these users to access only `Status`.
+  
+### Naming Conventions
+
+There are two naming decisions here to make:
+* Managed resource CR name
+* Identifier string of the external resource in the cloud provider.
```

Comment:
I see your point. Only in `Status` doesn't make sense either since we want to be able to reconstruct `Status` when it's lost. Having it in `Spec` doesn't make sense since it'd mean that a field is made available in _desired state_ but user can never actually desire it, that's already covered in _what goes to spec_ section

Annotation makes total sense in this case, also it's not an unfamiliar concept to indicate external config as annotation https://kubernetes.io/docs/concepts/services-networking/service/#ssl-support-on-aws

I couldn't see where you mention the case where the name should be chosen by us (claim controller case) and metadata name format doesn't comply with the provider's style, I guess you're onboard with having the developer write a name generator for that specific resource. This would also go to  `crossplane.io/external-name-override`.

In case where user creates the resource and it doesn't comply, I don't think we should take action other than showing the error.

I'll change the section as you suggested.

Reply:
How about we have the annotation in _all_ cases? In happy case, `Create` function of the given resource can assign the name in metadata to the annotation field's value so that we can make it easier for clients that wants to access the external name by pointing to only an annotation key instead of having them consider this alternation logic. Though this logic is set once per each resource type, so, it's not too much to ask them to be careful per resource type.